Sources/StreamChatUI/ChatMessageList/ChatMessageListVC.swift

```diff
@@ -244,11 +246,24 @@ open class ChatMessageListVC:
     }
 
     open func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
-        if channelController.state == .remoteDataFetched && indexPath.row + 1 >= tableView.numberOfRows(inSection: 0) - 5 {
-            channelController.loadPreviousMessages()
+        if channelController.state != .remoteDataFetched {
+            return
+        }
+
+        if indexPath.row < tableView.numberOfRows(inSection: 0) - 10 {
+            return
+        }
+
+        if _loadingPreviousMessages.compareAndSwap(old: false, new: true) {
+            channelController.loadPreviousMessages(completion: { [weak self] _ in
+                guard let self = self else {
+                    return
+                }
+                self.loadingPreviousMessages = false
+            })
```

Comment:
We don't need `loadingPreviousMessages` to be atomic. `willDisplay` function is always called on the main thread which is a serial thread, so there won't be any concurrent issues in this case because we can guarantee this logic is always called in the same thread. 

One problem I can see with this approach is that if the network is bad, or the user is scrolling too fast, there's a chance we might skip one loading request, since you will only be able to load previous messages one at the time. There's no reason for having only one loading request per pagination. If we do this, the last loading request will always impact the next one if the network conditions are not the best.